### PR TITLE
Add additional fields to account key added event

### LINKED ISSUE #3146: emit account key information in AccountKeyAdded event
### Issue to be solved

the old `AccountKeyAddedFromByteArrayEventType` event contained a publickey field which is a byte encoded value of `runtimeAccountPublicKeyWrapper` [injected dependency defined here](https://github.com/onflow/flow-go/blob/master/model/flow/account_encoder.go#L36) With this event payload one is able to identify all information of the public key that was added to this account except for index. 
the new `AccountKeyAddedFromPublicKeyEventType` event only contains the raw public key itself with no additional account key information. 

### Suggested Solution

emit the following information in `AccountKeyAddedFromPublicKeyEventType`:
- account address
- public key raw value
- sig algo
- sign algo
- weight
- keyIndex

### PR DESCRIPTION
**Pull request by rrrkren**
Closes #3146

## Description

adding the following information to `flow.AccountKeyAdded` event:
- hash algorithm
- weight
- key index

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by rrrkren** [comment]
@turbolent I did not make the changes you requested wrt `AccountKeyRemoved` because I was not sure what's the safest way to remove an event. Also the `AccountKeyRemoved` currently uses a field called `publicKey` to indicate the index of the publicKey.... which clashes with the `publicKey` field we have in `AccountKeyAdded`. So the consistency game is already off.....

File: runtime/stdlib/flow.go [hunk 1]
```diff
@@ -209,6 +209,21 @@ var AccountEventContractParameter = sema.Parameter{
 	TypeAnnotation: sema.StringTypeAnnotation,
 }
 
+var AccountEventHashAlgorithmParameter = sema.Parameter{
+	Identifier:     "hashAlgorithm",
+	TypeAnnotation: sema.UInt8TypeAnnotation,
```
Comment (rrrkren) [inline runtime/stdlib/flow.go:214]: not sure if this is the correctly to identify a enum

Comment (turbolent) [reply runtime/stdlib/flow.go:214]: Enums are essentially structs with an `rawValue` field. It's probably better to emit an event with more type information, i.e.

```suggestion
	TypeAnnotation: sema.HashAlgorithmTypeAnnotation,
```

The code in `newAccountKeysAddFunction` and in the test already is correct, only the type information here is wrong.

File: runtime/stdlib/flow.go [hunk 2]
```diff
@@ -209,6 +209,21 @@ var AccountEventContractParameter = sema.Parameter{
...
+}
+
+var AccountEventKeyWeightParameter = sema.Parameter{
+	Identifier:     "weight",
+	TypeAnnotation: sema.IntTypeAnnotation,
```
Comment (turbolent) [inline runtime/stdlib/flow.go:219]: It's unfortunate that we used `UFix64` for the weight in the Cadence API, in `AccountKey` and in `keys.add`, but under the hood it's just an `int`.

Still, to maybe be consistent, use `UFix64` here.

File: runtime/stdlib/account.go [hunk 3]
```diff
@@ -610,7 +610,8 @@ func newAccountKeysAddFunction(
 				panic(err)
 			}
 
-			hashAlgo := NewHashAlgorithmFromValue(inter, locationRange, invocation.Arguments[1])
+			hashAlgoValue := invocation.Arguments[1]
+			hashAlgo := NewHashAlgorithmFromValue(inter, locationRange, hashAlgoValue)
 			weightValue, ok := invocation.Arguments[2].(interpreter.UFix64Value)
```
Comment (turbolent) [inline runtime/stdlib/account.go:615]: While at it, maybe separate the handling of the arguments a bit more:

```suggestion

			weightValue, ok := invocation.Arguments[2].(interpreter.UFix64Value)
```

**Review by turbolent — COMMENTED** [review]
Nice! Just a couple minor things

**Comment by turbolent** [comment]
> @turbolent I did not make the changes you requested wrt `AccountKeyRemoved` because I was not sure what's the safest way to remove an event. Also the `AccountKeyRemoved` currently uses a field called `publicKey` to indicate the index of the publicKey.... which clashes with the `publicKey` field we have in `AccountKeyAdded`. So the consistency game is already off.....

Yeah, looks like the definition of `AccountKeyRemoved` would also need some clean up before adding more information, `publicKey` is not a public key, but an index, and it is the index of the account key.

**Comment by turbolent** [comment]
@rrrkren once this is merged, could you please also update the docs in https://github.com/onflow/cadence-lang.org/tree/main/versioned_docs/version-1.0 ?

File: runtime/account_test.go [hunk 4]
```diff
@@ -507,19 +515,42 @@ func TestRuntimeAuthAccountKeysAdd(t *testing.T) {
...
+	// key weight
+	key0Weight, err := cadence.NewUFix64("100.0")
+	require.NoError(t, err)
+	key1Weight, err := cadence.NewUFix64("0.0")
+	require.NoError(t, err)
+
+	assert.EqualValues(t, key0Weight, key0AddedEvent.Fields[2])
+	assert.EqualValues(t, sema.UFix64TypeName, key0AddedEvent.Fields[2].Type().ID())
+	assert.EqualValues(t, key1Weight, key1AddedEvent.Fields[2])
+
```
Comment (turbolent) [inline runtime/account_test.go:544]: Also add an assertion for the type of the third field of the second key

File: runtime/account_test.go [hunk 5]
```diff
@@ -507,19 +515,42 @@ func TestRuntimeAuthAccountKeysAdd(t *testing.T) {
...
+	// key hash algorithm
+	assert.EqualValues(t, sema.HashAlgorithmSHA3_256, key0AddedEvent.Fields[3].(cadence.Enum).Fields[0])
+	assert.EqualValues(t, sema.HashAlgorithmTypeName, key0AddedEvent.Fields[3].Type().ID())
+	assert.EqualValues(t, sema.HashAlgorithmSHA2_256, key1AddedEvent.Fields[3].(cadence.Enum).Fields[0])
+
```
Comment (turbolent) [inline runtime/account_test.go:549]: Also add an assertion for the type of the fourth field of the second key

File: runtime/account_test.go [hunk 6]
```diff
@@ -507,19 +515,42 @@ func TestRuntimeAuthAccountKeysAdd(t *testing.T) {
...
+	// key index
+	assert.EqualValues(t, cadence.NewInt(0), key0AddedEvent.Fields[4])
+	assert.EqualValues(t, sema.IntTypeName, key0AddedEvent.Fields[4].Type().ID())
+	assert.EqualValues(t, cadence.NewInt(1), key1AddedEvent.Fields[4])
 }
```
Comment (turbolent) [inline runtime/account_test.go:554]: Also add an assertion for type of the the fifth field of the second key

File: runtime/account_test.go [hunk 7]
```diff
@@ -507,19 +515,42 @@ func TestRuntimeAuthAccountKeysAdd(t *testing.T) {
...
+		key0AddedEvent.Type().ID(),
 	)
+
+	// key weight
+	key0Weight, err := cadence.NewUFix64("100.0")
+	require.NoError(t, err)
+	key1Weight, err := cadence.NewUFix64("0.0")
+	require.NoError(t, err)
+
+	assert.EqualValues(t, key0Weight, key0AddedEvent.Fields[2])
```
Comment (turbolent) [inline runtime/account_test.go:541]: Here an below: Use `assert.Equal` to also assert types are correct

File: runtime/account_test.go [hunk 8]
```diff
@@ -507,19 +515,42 @@ func TestRuntimeAuthAccountKeysAdd(t *testing.T) {
...
 
+	key0AddedEvent := storage.events[1]
+	key1AddedEvent := storage.events[2]
+
 	assert.EqualValues(t,
 		stdlib.AccountKeyAddedFromPublicKeyEventType.ID(),
-		storage.events[1].Type().ID(),
+		key0AddedEvent.Type().ID(),
 	)
+
```
Comment (turbolent) [inline runtime/account_test.go:534]: While at it, could you maybe please also add assertions for the first and second field?

**Review by turbolent — COMMENTED** [review]
Looks good!

Just a couple small nits for the tests

**Review by turbolent — APPROVED** [review]

**Comment by turbolent** [comment]
@onflow/cadence PTAL 🙏

**Review by SupunS — APPROVED** [review]

**Pull request merged**